addon/components/polaris-data-table.js

```diff
@@ -1,525 +1,490 @@
-      );
+  handleResize() {
+    // This is needed to replicate the React implementation's `@debounce` decorator.
+    this.debounceTask('debouncedHandleResize', 0);
+  }
 
-      if (table) {
-        let rows = Array.from(table.getElementsByTagName('tr'));
+  debouncedHandleResize() {
+    let { footerContent, truncate, table, scrollContainer } = this;
+
+    let collapsed = false;
+
+    if (table && scrollContainer) {
+      collapsed = table.scrollWidth > scrollContainer.clientWidth;
+      scrollContainer.scrollLeft = 0;
+    }
+
+    this.setProperties(
+      assign(
+        {
+          collapsed,
+          heights: [],
+        },
+        this.calculateColumnVisibilityData(collapsed)
+      )
+    );
+
+    if (footerContent || !truncate) {
+      scheduleOnce('afterRender', this, this.setHeightsAndScrollPosition);
+    }
+  }
 
-        if (!truncate) {
-          return (heights = rows.map((row) => {
-            let fixedCell = row.hasChildNodes() && row.children[0];
-            return Math.max(row.clientHeight, fixedCell.clientHeight);
-          }));
-        }
+  scrollListener() {
+    if (this.get('isDestroying') || this.get('isDestroyed')) {
+      return;
+    }
 
-        if (footerContent) {
-          let footerCellHeight =
-            rows[rows.length - 1].childNodes[0].clientHeight;
-          heights = [footerCellHeight];
-        }
+    this.setProperties(
+      this.calculateColumnVisibilityData(this.get('collapsed'))
+    );
+  }
 
-        return heights;
-      }
-    },
+  tallestCellHeights() {
+    let { footerContent, truncate, heights, table } = this;
 
-    addEventHandlers() {
-      this.addEventListener(window, 'resize', this.handleResize);
-      this.addEventListener(window, 'scroll', this.scrollListener, {
-        capture: true,
-      });
-    },
+    if (table) {
+      let rows = Array.from(table.getElementsByTagName('tr'));
 
-    init() {
-      this._super(...arguments);
+      if (!truncate) {
+        return (heights = rows.map((row) => {
+          let fixedCell = row.hasChildNodes() && row.children[0];
+          return Math.max(row.clientHeight, fixedCell.clientHeight);
+        }));
+      }
 
-      this.setProperties({
-        columnVisibilityData: [],
-        heights: [],
-        preservedScrollPosition: {},
-      });
-    },
+      if (footerContent) {
+        let footerCellHeight = rows[rows.length - 1].childNodes[0].clientHeight;
+        heights = [footerCellHeight];
+      }
 
-    didInsertElement() {
-      this._super(...arguments);
+      return heights;
+    }
+  }
 
-      this.handleResize();
+  addEventHandlers() {
+    this.addEventListener(window, 'resize', this.handleResize);
+    this.addEventListener(window, 'scroll', this.scrollListener, {
+      capture: true,
+    });
+  }
 
-      this.addEventHandlers();
-    },
+  init() {
+    super.init(...arguments);
 
-    didUpdateAttrs() {
-      this._super(...arguments);
+    this.setProperties({
```

Comment:
let's move these initializations where we declare the properties at the top of the file....it's now safe to do that 🙌